src/appfl/agent/server.py

```diff
@@ -0,0 +1,371 @@
+
+    def server_terminated(self):
+        """Indicate whether the server can be terminated from listening to the clients."""
+        if not hasattr(self, "closed_clients"):
+            return False
+        num_clients = (
+            self.server_agent_config.server_configs.num_clients if 
+            hasattr(self.server_agent_config.server_configs, "num_clients") else
+            self.server_agent_config.server_configs.scheduler_kwargs.num_clients if
+            hasattr(self.server_agent_config.server_configs.scheduler_kwargs, "num_clients") else
+            self.server_agent_config.server_configs.aggregator_kwargs.num_clients
+        )
+        with self._close_connection_lock:
+            terminated = len(self.closed_clients) >= num_clients
+        if terminated:
+            self.clean_up()
+        return terminated
+    
+    def clean_up(self) -> None:
+        """
+        Nececessary clean-up operations.
+        No need to call this method if using `server_terminated` to check the termination status.
+        """
+        if not hasattr(self, "cleaned"):
+            self.cleaned = False
+        if not self.cleaned:
+            self.cleaned = True
+            if hasattr(self.scheduler, "clean_up"):
+                self.scheduler.clean_up()
+
+    def _create_logger(self) -> None:
+        kwargs = {}
+        if hasattr(self.server_agent_config.server_configs, "logging_output_dirname"):
+            kwargs["file_dir"] = self.server_agent_config.server_configs.logging_output_dirname
+        if hasattr(self.server_agent_config.server_configs, "logging_output_filename"):
+            kwargs["file_name"] = self.server_agent_config.server_configs.logging_output_filename
+        self.logger = ServerAgentFileLogger(**kwargs)
+
+    def _load_model(self) -> None:
+        """
+        Load model from the definition file, and read the source code of the model for sendind to the client.
+        User can overwrite this method to load the model from other sources.
+        """
+        self._set_seed()
+        model_configs = (
+            self.server_agent_config.client_configs.model_configs 
+            if hasattr(self.server_agent_config.client_configs, "model_configs") 
+            else self.server_agent_config.server_configs.model_configs
+        )
+        self.model = create_instance_from_file(
+            model_configs.model_path,
+            model_configs.model_name,
+            **(model_configs.model_kwargs if hasattr(model_configs, "model_kwargs") else {})
+        )
+        # load the model source file and delete model path
+        if hasattr(self.server_agent_config.client_configs, "model_configs"):
+            with open(model_configs.model_path, 'r') as f:
+                self.server_agent_config.client_configs.model_configs.model_source = f.read()
+            del self.server_agent_config.client_configs.model_configs.model_path
+
+    def _load_loss(self) -> None:
```

Comment:
To my understanding, this needs users to create a separate file that defines a loss function class. 
- Do users need to follow a certain form for the class? In example, the loss function class defines `__init__` and `forward`. Is the `forward` method required to implement?
- Is there a way for users to simply indicate pytorch function?

Reply:
As shown [here](https://github.com/APPFL/APPFL/blob/zilinghan/grpc-update/docs/users/user_loss.rst), user need to define loss as a `torch.nn.Module` with a defined forward function to compute the loss. Compared with a simple pytorch function, this gives user some flexibility to have some additional arguments when intializing the loss function, e.g., `CrossEntroyLoss(reduction='mean')`.

Reply:
Can you docuement this, if not already done?

Reply:
Here the the document for this https://github.com/APPFL/APPFL/blob/zilinghan/grpc-update/docs/users/user_loss.rst